tensorflow_addons/seq2seq/README.md

```diff
@@ -0,0 +1,184 @@
+The code was originally written in tensorflow.contrib.seq2seq, and has been updated to work with
+TF 2.0 API. The API has been reworked to get rid of deprecated TF APIs (eg, using variable
+scope to create variable, etc), and also meet the 2.0 API sytle (more object-oriented and use keras
+layers). With that, the user side code need to be slightly updated to use the new API. Please see
+examples below:
+
+### Basic Decoder
+``` python
+# TF 1.x, old style
+
+# Build RNN cell
+encoder_cell = tf.nn.rnn_cell.BasicLSTMCell(num_units)
+
+# Run Dynamic RNN
+#   encoder_outputs: [max_time, batch_size, num_units]
+#   encoder_state: [batch_size, num_units]
+encoder_outputs, encoder_state = tf.nn.dynamic_rnn(
+    encoder_cell, encoder_emb_inp,
+    sequence_length=source_sequence_length)
+
+# Build RNN cell
+decoder_cell = tf.nn.rnn_cell.BasicLSTMCell(num_units)
+
+# Helper
+helper = tf.contrib.seq2seq.TrainingHelper(
+    decoder_emb_inp, decoder_lengths)
+# Decoder
+decoder = tf.contrib.seq2seq.BasicDecoder(
+    decoder_cell, helper, encoder_state,
+    output_layer=projection_layer)
+# Dynamic decoding
+outputs, _ = tf.contrib.seq2seq.dynamic_decode(decoder, ...)
+logits = outputs.rnn_output
+```
+
+``` python
+import tensorflow_addons as tfa
+
+# TF 2.0, new style
+
+# Build RNN
+#   encoder_outputs: [max_time, batch_size, num_units]
+#   encoder_state: [batch_size, num_units]
+encoder = tf.keras.layers.LSTM(num_units)
+encoder_outputs, state_h, state_c = encoder(encoder_emb_inp)
+encoder_state = [state_h, state_c]
+
+# Sampler
+sampler = tfa.seq2seq.TrainingSampler()
+
+# Decoder
+decoder_cell = tf.keras.layers.LSTMCell(num_units)
+decoder = tfa.seq2seq.BasicDecoder(
+    decoder_cell, sampler, output_layer=projection_layer))
+
+outputs, _ = decoder(
+    decoder_emb_inp,
+    initial_state=encoder_state,
+    sequence_length=decoder_lengths)
+logits = outputs.rnn_output
+```
+
+Note that the major difference here are:
+
+1. Both encoder and decoder are objects now, and all the metadata can be accessed, eg,
+   `encoder.weights`, etc.
+1. All the tensor inputs are feed to encoder/decoder by calling it, instead of feedin when construct
+   the instance. This allows the same instance to be reused if needed, just call it with other input
+   tensors.
+1. Helper has been renamed to Sampler since it describer its behavior/usage better. There is a
+   one-to-one mapping between existing Helper and new Sampler. Sampler is also a keras layer, which
+   takes input tensors at `call()` instead of `__init__()`.
+
+
+### Attention
+``` python
+# TF 1.x, old style
+attention_mechanism = tf.contrib.seq2seq.LuongAttention(
+    num_units,
+    encoder_state,
+    memory_sequence_length=encoder_sequence_length)
+
+decoder_cell = tf.contrib.seq2seq.AttentionWrapper(
+    decoder_cell, attention_mechanism,
+    attention_layer_size=num_units)
+```
+
+``` python
+import tensorflow_addons as tfa
+# TF 2.0, new style
+attention_mechanism = tfa.seq2seq.LuongAttention(
```

Comment:
That RFC was only covering the DNN's case. During the internal review, we explicitly exclude the RNN from it since the RNN is too complicated.

Reply:
Yes the RFC was for Dense, but there was mention of work on the RNN. It's not super important for this merge, but was curious to know what the landscape looks like going forward.
> tf.contrib.seq2seq.LuongAttention and tf.contrib.seq2seq.BahdanauAttention are implementations of dot-product (Luong) and additive (Bahdanau) Attention respectively for RNN in Tensorflow. There is ongoing work to implement those as Keras layers

Reply:
Ah, the LuongAttention and BahdanauAttention has already been reworked as a keras layer, and also, the luong and bahdanau function has been standardized so that it can be imported/used by other similar use case. I guess the DNN attention layer could use those use those methods/layers, or we can move those functions to core and use them from there.